### PR TITLE
Options refactor

### PR DESCRIPTION
**Pull request by waltsims**
This PR takes on the task of refactoring all of the options logic for k-Wave.

still left to do:
- [ ] simplify PML logic
- [ ] better option string generation
- [ ] simplify pml alpha logic

### CODE REVIEW

**Comment by waltsims** [comment]
> This PR takes on the task of refactoring all of the options logic for k-Wave.
> 
> still left to do:
> 
>     * [ ]  simplify PML logic
> 
>     * [ ]  better option string generation
> 
>     * [ ]  simplify pml alpha logic

These will be individual PRs

File: tests/test_cpp_running_simulations.py [hunk 1]
```diff
@@ -35,14 +35,14 @@ def test_cpp_running_simulations():
     example_number = 1
 
     # input and output filenames (these must have the .h5 extension)
-    input_filename  = 'example_input.h5'
+    input_filename = 'example'
```
Comment (faridyagubbayli) [inline tests/test_cpp_running_simulations.py:38]: why did you separate the suffix?

File: tests/test_ivp_3D_simulation.py [hunk 2]
```diff
@@ -70,13 +70,16 @@ def test_ivp_3D_simulation():
     sensor = kSensor(sensor_mask)
 
     # input arguments
-    input_filename  = f'example_input.h5'
-    input_file_full_path = os.path.join(pathname, input_filename)
+    input_filename = f'example_ivp_3D'
```
Comment (faridyagubbayli) [inline tests/test_ivp_3D_simulation.py:73]: Same here?

File: kwave/options.py [hunk 3]
```diff
@@ -191,131 +251,37 @@ def init(kgrid, elastic_code: bool, axisymmetric: bool, **kwargs):
...
-        if options.pml_auto:
-            if options.pml_inside:
-                raise NotImplementedError("''PMLSize'' set to ''auto'' is only supported with ''PMLInside'' set to false.")
-            else:
-                # get optimal pml size
-                if axisymmetric:
-                    pml_size_temp = get_optimal_pml_size(kgrid, options.pml_search_range, options.radial_symmetry[:4])
-                else:
-                    pml_size_temp = get_optimal_pml_size(kgrid, options.pml_search_range)
+                assert kgrid.dim == 1 and not options.elastic_code, "Optional input ''UseFD'' only supported in 1D."
```
Comment (faridyagubbayli) [inline kwave/options.py:269]:
```suggestion
                assert kgrid.dim == 1 and not options.elastic_code, "Optional input ''use_fd'' only supported in 1D."
```

File: kwave/options.py [hunk 4]
```diff
@@ -1,182 +1,242 @@
...
-
-            elif key == 'PMLAlpha':
+            if key == 'pml_alpha':
                 # check input is correct size
                 val = np.atleast_1d(val)
                 if val.size > kgrid.dim:
                     if kgrid.dim > 1:
-                        raise ValueError(f"Optional input ''PMLAlpha'' must be a 1 or {kgrid.dim} element numerical array.")
+                        raise ValueError(
+                            f"Optional input ''PMLAlpha'' must be a 1 or {kgrid.dim} element numerical array.")
```
Comment (faridyagubbayli) [inline kwave/options.py:239]:
```suggestion
                            f"Optional input ''pml_alpha'' must be a 1 or {kgrid.dim} element numerical array.")
```

File: kwave/options.py [hunk 5]
```diff
@@ -1,182 +1,242 @@
...
 
-        options = SimulationOptions()
+        if options.pml_size is not None or not isinstance(options.pml_size, bool):
+            options.pml_size = np.atleast_1d(options.pml_size)
+            if len(options.pml_size) > kgrid.dim:
+                if kgrid.dim > 1:
+                    raise ValueError(
+                        f"Optional input ''PMLSize'' must be a 1 or {kgrid.dim} element numerical array.")
+                else:
+                    raise ValueError(f"Optional input ''PMLSize'' must be a single numerical value.")
```
Comment (faridyagubbayli) [inline kwave/options.py:207]:
```suggestion
                        f"Optional input ''pml_size'' must be a 1 or {kgrid.dim} element numerical array.")
                else:
                    raise ValueError(f"Optional input ''pml_size'' must be a single numerical value.")
```

File: kwave/options.py [hunk 6]
```diff
@@ -1,182 +1,242 @@
...
+        for key, val in boolean_inputs.items():
+            assert isinstance(val, bool), f"Optional input ''{key}'' must be Boolean."
+
+        assert self.radial_symmetry in ['WSWA', 'WSWS', 'WSWA-FFT', 'WSWS-FFT'], \
+            "Optional input ''RadialSymmetry'' must be set to ''WSWA'', ''WSWS'', ''WSWA-FFT'', ''WSWS-FFT''."
+
+        # automatically assign the PML size to give small prime factors
+        if self.pml_auto and self.pml_inside:
+            raise NotImplementedError(
+                "''PMLSize'' set to ''auto'' is only supported with ''PMLInside'' set to false.")
```
Comment (faridyagubbayli) [inline kwave/options.py:129]:
```suggestion
                "''pml_size'' set to ''auto'' is only supported with ''pml_inside'' set to false.")
```

File: kwave/options.py [hunk 7]
```diff
@@ -1,182 +1,242 @@
...
+
         # load the HDF5 literals (for the default compression level)
         h5_literals = get_h5_literals()
         self.hdf_compression_level = h5_literals.HDF_COMPRESSION_LEVEL
+        # check value is an integer between 0 and 9
+        assert isinstance(self.hdf_compression_level, int) and 0 <= self.hdf_compression_level <= 9, \
+            "Optional input ''HDFCompressionLevel'' must be an integer between 0 and 9."
+
+        assert np.isscalar(self.multi_axial_PML_ratio) and self.multi_axial_PML_ratio >= 0, \
+            "Optional input ''MultiAxialPMLRatio'' must be a single positive value."
```
Comment (faridyagubbayli) [inline kwave/options.py:106]:
```suggestion
            "Optional input ''multi_axial_PML_ratio'' must be a single positive value."
```

File: kwave/options.py [hunk 8]
```diff
@@ -1,182 +1,242 @@
...
+
+        assert np.isscalar(self.stream_to_disk) or isinstance(self.stream_to_disk, bool), \
+            "Optional input ''StreamToDisk'' must be a single scalar or Boolean value."
```
Comment (faridyagubbayli) [inline kwave/options.py:109]:
```suggestion
            "Optional input ''stream_to_disk'' must be a single scalar or Boolean value."
```

File: kwave/options.py [hunk 9]
```diff
@@ -1,182 +1,242 @@
...
+
+        if self.data_cast == 'double':
+            self.data_cast = 'off'
+
         # load the HDF5 literals (for the default compression level)
         h5_literals = get_h5_literals()
         self.hdf_compression_level = h5_literals.HDF_COMPRESSION_LEVEL
+        # check value is an integer between 0 and 9
+        assert isinstance(self.hdf_compression_level, int) and 0 <= self.hdf_compression_level <= 9, \
+            "Optional input ''HDFCompressionLevel'' must be an integer between 0 and 9."
```
Comment (faridyagubbayli) [inline kwave/options.py:103]:
```suggestion
            "Optional input ''hdf_compression_level'' must be an integer between 0 and 9."
```

File: kwave/options.py [hunk 10]
```diff
@@ -1,182 +1,242 @@
...
+    pml_y_alpha: Optional[float] = None
+    pml_z_alpha: Optional[float] = None
+    pml_size: Optional[List[int]] = None
+    pml_x_size: Optional[int] = None
+    pml_y_size: Optional[int] = None
+    pml_z_size: Optional[int] = None
 
     def __post_init__(self):
+        assert self.cartesian_interp in ['linear', 'nearest'], \
+            "Optional input ''CartInterp'' must be set to ''linear'' or ''nearest''."
```
Comment (faridyagubbayli) [inline kwave/options.py:88]:
```suggestion
            "Optional input ''cartesian_interp'' must be set to ''linear'' or ''nearest''."
```

**Review by faridyagubbayli — APPROVED** [review]
LGTM! Left a few minor comments.

File: tests/test_cpp_running_simulations.py (hunk 1 above)
Comment (waltsims) [reply tests/test_cpp_running_simulations.py:38]: You're right. this doesn't make sense...

File: tests/test_ivp_3D_simulation.py (hunk 2 above)
Comment (waltsims) [reply tests/test_ivp_3D_simulation.py]: addressed in [d876b13](https://github.com/waltsims/k-wave-python/pull/64/commits/d876b13e871b4a1860f672ecc3dd402383e218e9)

File: kwave/kspaceFirstOrder.py [hunk 11]
```diff
@@ -59,69 +59,65 @@ def wrapper(**kwargs):
...
                     data_path = data_path + os.path.sep
-                del kwargs['DataPath']
             else:
                 # set default path
                 data_path = gettempdir()
 
             # check for a user defined name for the input and output files
-            if 'DataName' in kwargs.keys():
-                name_prefix = kwargs['DataName']
+            if 'data_name' in kwargs.keys():
```
Comment (waltsims) [inline kwave/kspaceFirstOrder.py:117]: This is likely dead code...

File: kwave/options.py [hunk 12]
```diff
@@ -1,182 +1,230 @@
...
+    pml_size: Optional[List[int]] = None
+    pml_x_size: Optional[int] = None
+    pml_y_size: Optional[int] = None
+    pml_z_size: Optional[int] = None
 
     def __post_init__(self):
+        assert self.cartesian_interp in ['linear', 'nearest'], \
+            "Optional input ''cartesian_interp'' must be set to ''linear'' or ''nearest''."
+
+        assert isinstance(self.data_cast, str), "Optional input ''DataCast'' must be a string."
```
Comment (waltsims) [inline kwave/options.py:89]: DataCast -> data_cast

File: kwave/options.py [hunk 13]
```diff
@@ -1,182 +1,230 @@
...
+            if key == 'pml_alpha':
                 # check input is correct size
                 val = np.atleast_1d(val)
                 if val.size > kgrid.dim:
                     if kgrid.dim > 1:
-                        raise ValueError(f"Optional input ''PMLAlpha'' must be a 1 or {kgrid.dim} element numerical array.")
+                        raise ValueError(
+                            f"Optional input ''pml_alpha'' must be a 1 or {kgrid.dim} element numerical array.")
                     else:
                         raise ValueError(f"Optional input ''PMLAlpha'' must be a single numerical value.")
```
Comment (waltsims) [inline kwave/options.py:229]: PMLAlpha -> pml_alpha

File: kwave/kspaceFirstOrder.py (hunk 11 above)
Comment (waltsims) [reply kwave/kspaceFirstOrder.py:117]: we can look at this later...

**Pull request merged**